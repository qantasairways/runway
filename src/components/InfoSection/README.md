```js
<InfoSection
  content={
    <div>
      Info Content <a href="https://www.google.com.au">with link</a>
    </div>
  }
/>
```
