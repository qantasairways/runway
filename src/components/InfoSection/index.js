import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';

import InfoIcon from '../../icons/InfoIcon';
import { colours } from '../../theme/airways';

const iconSize = '16.8px';

const getContainerStyles = ({ isMultiLine, height, width }) => ({
  height,
  width,
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  marginRight: '10px',
  ...(isMultiLine && {
    position: 'relative'
  })
});

const getIconStyles = ({ isMultiLine }) => ({
  fill: colours.darkerGrey,
  marginRight: isMultiLine ? '0px' : '11px'
});

const getIconContainerStyles = ({ isMultiLine, lines }) => ({
  display: 'flex',
  ...(isMultiLine && {
    position: 'absolute',
    top: lines > 2 ? '2px' : '1px',
    left: '0'
  })
});

const getContentStyles = ({ isMultiLine }) => ({
  color: colours.darkerGrey,
  lineHeight: '1.21',
  fontSize: '14px',
  fontFamily: 'Ciutadella',
  ...(isMultiLine && { marginLeft: '27.8px' })
});

class InfoSection extends React.Component {
  state = { lines: 1 };

  setContentRef = el => {
    this.contentNode = el;
  };

  componentDidMount = () => {
    this.updateLineCount();
    this.debouncedUpdateLineCount = debounce(() => {
      this.updateLineCount();
    }, 200);
    window.addEventListener('resize', this.debouncedUpdateLineCount);
  };

  componentWillUnmount = () => {
    window.removeEventListener('resize', this.debouncedUpdateLineCount);
  };

  isMultiLine = () => this.state.lines > 1;

  updateLineCount = () => {
    const el = this.contentNode;
    if (el) {
      const elHeight = el.offsetHeight;
      const contentStyles = getContentStyles({
        isMultiLine: this.isMultiLine()
      });
      const lineHeight =
        Number(contentStyles.lineHeight) *
        Number(contentStyles.fontSize.replace('px', ''));
      const lines = elHeight / lineHeight + 1;
      this.setState({ lines: parseInt(lines, 10) });
    }
  };

  render = () => {
    const { content, height, width } = this.props;
    const isMultiLine = this.isMultiLine();
    return (
      <div css={getContainerStyles({ height, width, isMultiLine })}>
        <span
          css={getIconContainerStyles({ isMultiLine, lines: this.state.lines })}
        >
          <InfoIcon
            css={getIconStyles({ isMultiLine })}
            height={iconSize}
            width={iconSize}
          />
        </span>
        <span
          // eslint-disable-next-line no-return-assign
          ref={this.setContentRef}
          className="runway-info-section__text-content"
          css={getContentStyles({ isMultiLine })}
        >
          {content}
        </span>
      </div>
    );
  };
}

InfoSection.propTypes = {
  content: PropTypes.node.isRequired,
  height: PropTypes.string,
  width: PropTypes.string
};

InfoSection.defaultProps = {
  height: 'auto',
  width: '100%;'
};

export default InfoSection;
