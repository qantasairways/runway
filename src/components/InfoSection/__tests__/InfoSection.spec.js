import React from 'react';
import { mount } from 'enzyme';
import InfoSection from '..';

jest.mock('../../../icons/InfoIcon', () => () => <span>Mock Info Icon</span>);

describe('InfoSection', () => {
  it('renders correctly with props', () => {
    const requiredProps = {
      content: 'mock content',
      height: '100px',
      width: '200px'
    };
    const component = mount(<InfoSection {...requiredProps} />);
    expect(component).toMatchSnapshot();
  });
});
