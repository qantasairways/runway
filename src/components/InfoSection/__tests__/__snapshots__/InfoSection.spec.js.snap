// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoSection renders correctly with props 1`] = `
.emotion-3 {
  height: 100px;
  width: 200px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 10px;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0 {
  fill: #323232;
  margin-right: 11px;
}

.emotion-2 {
  color: #323232;
  line-height: 1.21;
  font-size: 14px;
  font-family: Ciutadella;
}

<div
  className="emotion-3"
>
  <span
    className="emotion-1"
  >
    <InfoIcon
      className="emotion-0"
      height="16.8px"
      width="16.8px"
    />
  </span>
  <span
    className="emotion-2 runway-info-section__text-content"
  >
    mock content
  </span>
</div>
`;
